packages/ruleset/src/functions/use-date-based-format.js

```diff
@@ -0,0 +1,359 @@
+ * date-based schemas. As it traverses the schemas, it compiles a list of
+ * potentially-relevant example values. This way, if an object schema defines
+ * its own example, which includes a value for a nested property that should
+ * be identified by the rule, we can track down the value once we reach the
+ * schema for said property. The logic will also gather any relevant parameter
+ * or media type examples that may be defined outside of the schema path.
+ *
+ * @param {object} s the schema to check
+ * @param {array} p the array of path segments indicating the "location" of the schema within the API definition
+ * @param {object} apidef the resolved API definition
+ * @returns an array containing the violations found or [] if no violations
+ */
+function checkForDateBasedFormat(s, p, apidef) {
+  const examples = [];
+  const propertyPath = [];
+
+  return validateSubschemas(s, p, (schema, path) => {
+    logger.debug(`${ruleId}: checking schema at location: ${path.join('.')}`);
+
+    // Check for any examples outside of the schema path - they may be in
+    // request bodies, response bodies, or parameters.
+    examples.push(...checkForIndirectExamples(path, apidef));
+
+    // We can look at primitive schemas directly but for objects and arrays,
+    // we need more specialized handling in case we need to find a particular
+    // property within their examples.
+    if (isObjectSchema(schema) || isArraySchema(schema)) {
+      // Maintain a running path to each schema that we look at. This will be
+      // used to determine where to look for a property value within an example
+      // object, relative to that example's location.
+      if (isSchemaProperty(path)) {
+        propertyPath.push(path.at(-1));
+      }
+
+      // Keep a running hierarchy of all examples we find as we look through
+      // the schemas in the API. Nested properties may only have an example
+      // value defined within a parent schema example.
+      if (schema.example) {
+        logger.debug(
+          `${ruleId}: adding example for schema at location: ${path.join('.')}`
+        );
+
+        examples.push({
+          example: schema.example,
+          examplePath: propertyPath.slice(), // Use a copy to prevent modification.
+        });
+      }
+
+      // Add sentinels for arrays/dictionaries to the running path,
+      // to assist the example-parsing logic later on. This must come
+      // after we push the example to the list.
+      if (isSchemaProperty(path)) {
+        if (isArraySchema(schema)) {
+          propertyPath.push('[]');
+        }
+
+        if (isDictionarySchema(schema)) {
+          propertyPath.push('{}');
+        }
+      }
+    }
+
+    // Use a slice (a copy) of the `propertyPath` array so that the
+    // invoked function can modify it without modifying the original.
+    return performValidation(
+      schema,
+      path,
+      apidef,
+      propertyPath.slice(),
```

Comment:
Personally, I think it is generally a bad practice to modify a parameter, so this should be done within the function. I also slightly prefer the clarity of `[... propertyPath]`

Reply:
Good call - will update 👍